File: analysis_cli/src/FrequencyResponseAnalyzer.cpp
#include <juce_dsp/juce_dsp.h>
#include <WDFilters/BandPassFilter.h>
#include <WDFilters/HighPassFilter.h>
#include <WDFilters/LowPassFilter.h>
#include <WDFilters/WDFilter.h>

#include <cmath>
#include <complex>
#include <iostream>
#include <memory>
#include <string>
#include <vector>

#include "Utils.h"

/**
 * @brief Calculate the frequency response of a filter
 * @param filter Pointer to the filter to analyze
 * @param sampleRate Sample rate in Hz
 * @param fftOrder FFT order (power of 2)
 * @return Tuple containing frequencies, magnitudes (in dB), and phases (in degrees)
 */
static std::tuple<std::vector<double>, std::vector<double>, std::vector<double>>
calculateFrequencyResponse(std::unique_ptr<WDFilter>& filter, double sampleRate, int fftOrder)
{
    const int      fftSize = 1 << fftOrder;
    juce::dsp::FFT fft(fftOrder);

    // Create buffer for FFT (2x size for complex output)
    std::vector<float> fftData(static_cast<size_t>(2 * fftSize), 0.0f);

    // Generate contiguous impulse response in the first half
    fftData[0] = 1.0f; // impulse
    for (int n = 0; n < fftSize; ++n)
        fftData[n] = filter->processSample(fftData[n]);

    // Perform FFT (with scaling)
    fft.performRealOnlyForwardTransform(fftData.data(), true);

    const int           numBins = fftSize / 2;
    std::vector<double> freq(numBins), magDb(numBins), phaseDeg(numBins);

    // Find maximum magnitude for normalization
    double maxMag = 0.0;
    for (int k = 0; k < numBins; ++k)
        maxMag = std::max(maxMag, static_cast<double>(std::hypot(fftData[2 * k], fftData[2 * k + 1])));

    // Calculate frequency response with phase unwrapping
    double prev = 0.0; // unwrap helper
    for (int k = 0; k < numBins; ++k)
    {
        const float re = fftData[2 * k];
        const float im = fftData[2 * k + 1];

        const double mag = std::hypot(re, im);
        double       ph  = std::atan2(im, re); // rad

        // Unwrap phase to ensure continuity
        double delta = ph - prev;
        if (delta > M_PI)
            ph -= 2 * M_PI;
        else if (delta < -M_PI)
            ph += 2 * M_PI;
        prev = ph;

        freq[k]     = k * sampleRate / fftSize;
        magDb[k]    = 20.0 * std::log10(mag / maxMag);
        phaseDeg[k] = ph * 180.0 / M_PI;
    }

    return {freq, magDb, phaseDeg};
}

int main([[maybe_unused]] int argc, [[maybe_unused]] char* argv[])
{
    // Define constants
    constexpr double sampleRate = 48000.0;
    constexpr double cutoffFreq = 1000.0;
    constexpr int    fftOrder   = 14; // 16384-point FFT

    // Create output directory
    fs::path outputDir = fs::current_path() / "frequency_responses";
    if (!utils::createDirectory(outputDir))
    {
        std::cerr << "Failed to create output directory" << std::endl;
        return 1;
    }

    std::cout << "Generating frequency response CSVs for all filter types..." << std::endl;
    std::cout << "Output directory: " << outputDir.string() << std::endl;

    // Generate responses for low pass filters (1st and 2nd order)
    {
        auto filter = WDFilter::create(WDFilter::Type::LowPass, WDFilter::Order::First);
        filter->prepare(sampleRate);
        filter->setCutoff(cutoffFreq);

        auto [frequencies, magnitudes, phases] = calculateFrequencyResponse(filter, sampleRate, fftOrder);
        std::string filename                   = utils::generateFilename("LowPass", 1, cutoffFreq);
        utils::writeCSV(outputDir / filename, frequencies, magnitudes, phases);
        std::cout << "Generated " << filename << std::endl;
    }

    {
        auto filter = WDFilter::create(WDFilter::Type::LowPass, WDFilter::Order::Second);
        filter->prepare(sampleRate);
        filter->setCutoff(cutoffFreq);

        auto [frequencies, magnitudes, phases] = calculateFrequencyResponse(filter, sampleRate, fftOrder);
        std::string filename                   = utils::generateFilename("LowPass", 2, cutoffFreq);
        utils::writeCSV(outputDir / filename, frequencies, magnitudes, phases);
        std::cout << "Generated " << filename << std::endl;
    }

    // Generate responses for high pass filters (1st and 2nd order)
    {
        auto filter = WDFilter::create(WDFilter::Type::HighPass, WDFilter::Order::First);
        filter->prepare(sampleRate);
        filter->setCutoff(cutoffFreq);

        auto [frequencies, magnitudes, phases] = calculateFrequencyResponse(filter, sampleRate, fftOrder);
        std::string filename                   = utils::generateFilename("HighPass", 1, cutoffFreq);
        utils::writeCSV(outputDir / filename, frequencies, magnitudes, phases);
        std::cout << "Generated " << filename << std::endl;
    }

    {
        auto filter = WDFilter::create(WDFilter::Type::HighPass, WDFilter::Order::Second);
        filter->prepare(sampleRate);
        filter->setCutoff(cutoffFreq);

        auto [frequencies, magnitudes, phases] = calculateFrequencyResponse(filter, sampleRate, fftOrder);
        std::string filename                   = utils::generateFilename("HighPass", 2, cutoffFreq);
        utils::writeCSV(outputDir / filename, frequencies, magnitudes, phases);
        std::cout << "Generated " << filename << std::endl;
    }

    // Generate responses for band pass filters (1st and 2nd order)
    {
        auto filter = WDFilter::create(WDFilter::Type::BandPass, WDFilter::Order::First);
        filter->prepare(sampleRate);
        filter->setCutoff(cutoffFreq);

        auto [frequencies, magnitudes, phases] = calculateFrequencyResponse(filter, sampleRate, fftOrder);
        std::string filename                   = utils::generateFilename("BandPass", 1, cutoffFreq);
        utils::writeCSV(outputDir / filename, frequencies, magnitudes, phases);
        std::cout << "Generated " << filename << std::endl;
    }

    {
        auto filter = WDFilter::create(WDFilter::Type::BandPass, WDFilter::Order::Second);
        filter->prepare(sampleRate);
        filter->setCutoff(cutoffFreq);

        auto [frequencies, magnitudes, phases] = calculateFrequencyResponse(filter, sampleRate, fftOrder);
        std::string filename                   = utils::generateFilename("BandPass", 2, cutoffFreq);
        utils::writeCSV(outputDir / filename, frequencies, magnitudes, phases);
        std::cout << "Generated " << filename << std::endl;
    }

    std::cout << "Frequency response analysis complete." << std::endl;

    return 0;
}

File: prototypes/src/frequency_response_analyzer.py
import csv
from pathlib import Path
from typing import List, Tuple, Union

import matplotlib.pyplot as plt
import numpy as np
from scipy import signal

from .rc_1st2ndorder_bandpass import RCBandPass1st, RCBandPass2nd
from .rc_highpass import RCHighPass
from .rc_lowpass import RCLowPass


def calculate_frequency_response(
    filter_instance: Union[RCLowPass, RCHighPass, RCBandPass1st, RCBandPass2nd],
    sample_rate: float,
    fft_order: int = 14,
) -> Tuple[np.ndarray, np.ndarray, np.ndarray]:
    """
    Calculate the frequency response of a filter using FFT.

    Parameters
    ----------
    filter_instance : Union[RCLowPass, RCHighPass, RCBandPass1st, RCBandPass2nd]
        The filter instance to analyze
    sample_rate : float
        Sample rate in Hz
    fft_order : int, optional
        FFT order (power of 2), by default 14 (16384-point FFT)

    Returns
    -------
    Tuple[np.ndarray, np.ndarray, np.ndarray]
        Arrays containing frequencies (Hz), magnitudes (dB), and phases (degrees)
    """
    fft_size = 1 << fft_order  # 2^fft_order

    # Create impulse signal
    impulse = np.zeros(fft_size)
    impulse[0] = 1.0

    # Process impulse through filter
    response = filter_instance.process_block(impulse)

    # Perform FFT
    spectrum = np.fft.rfft(response)
    frequencies = np.fft.rfftfreq(fft_size, 1 / sample_rate)

    # Calculate magnitude spectrum and convert to dB
    magnitudes = np.abs(spectrum)
    max_magnitude = np.max(magnitudes)
    magnitudes_db = 20 * np.log10(magnitudes / max_magnitude)

    # Calculate phase and unwrap
    # phases = np.angle(spectrum, deg=True)  # Get phase in degrees
    # phases = np.unwrap(phases)  # Unwrap phase to ensure continuity
    phases = np.unwrap(np.angle(spectrum)) * 180 / np.pi

    return frequencies, magnitudes_db, phases


def write_csv(
    filename: Path, frequencies: np.ndarray, magnitudes: np.ndarray, phases: np.ndarray
) -> None:
    """
    Write frequency response data to a CSV file.

    Parameters
    ----------
    filename : Path
        Output file path
    frequencies : np.ndarray
        Array of frequencies in Hz
    magnitudes : np.ndarray
        Array of magnitudes in dB (normalized to 0 dB peak)
    phases : np.ndarray
        Array of phases in degrees (unwrapped)
    """
    with open(filename, "w", newline="") as f:
        writer = csv.writer(f)
        writer.writerow(["frequency_hz", "magnitude_db", "phase_degrees"])
        for freq, mag, phase in zip(frequencies, magnitudes, phases):
            writer.writerow([f"{freq:.6f}", f"{mag:.6f}", f"{phase:.6f}"])


def generate_filename(filter_type: str, order: int, cutoff_freq: float) -> str:
    """
    Generate a filename for the frequency response CSV.

    Parameters
    ----------
    filter_type : str
        Type of filter (LowPass, HighPass, BandPass)
    order : int
        Filter order (1 or 2)
    cutoff_freq : float
        Cutoff frequency in Hz

    Returns
    -------
    str
        Generated filename
    """
    return f"pywdf_{filter_type}_order{order}_{int(cutoff_freq)}Hz.csv"


def main():
    # Define constants
    SAMPLE_RATE = 48000.0
    CUTOFF_FREQ = 1000.0
    FFT_ORDER = 14  # 16384-point FFT

    # Create output directory
    output_dir = Path("frequency_responses")
    output_dir.mkdir(exist_ok=True)

    print("Generating frequency response CSVs for all filter types...")
    print(f"Output directory: {output_dir.absolute()}")

    # Generate responses for low pass filters
    for order in [1, 2]:
        filter_instance = RCLowPass(SAMPLE_RATE, CUTOFF_FREQ)
        frequencies, magnitudes, phases = calculate_frequency_response(
            filter_instance, SAMPLE_RATE, FFT_ORDER
        )
        filename = generate_filename("LowPass", order, CUTOFF_FREQ)
        write_csv(output_dir / filename, frequencies, magnitudes, phases)
        print(f"Generated {filename}")

    # Generate responses for high pass filters
    for order in [1, 2]:
        filter_instance = RCHighPass(SAMPLE_RATE, CUTOFF_FREQ)
        frequencies, magnitudes, phases = calculate_frequency_response(
            filter_instance, SAMPLE_RATE, FFT_ORDER
        )
        filename = generate_filename("HighPass", order, CUTOFF_FREQ)
        write_csv(output_dir / filename, frequencies, magnitudes, phases)
        print(f"Generated {filename}")

    # Generate responses for band pass filters
    for order in [1, 2]:
        if order == 1:
            filter_instance = RCBandPass1st(SAMPLE_RATE, CUTOFF_FREQ)
        else:
            filter_instance = RCBandPass2nd(SAMPLE_RATE, CUTOFF_FREQ)

        frequencies, magnitudes, phases = calculate_frequency_response(
            filter_instance, SAMPLE_RATE, FFT_ORDER
        )
        filename = generate_filename("BandPass", order, CUTOFF_FREQ)
        write_csv(output_dir / filename, frequencies, magnitudes, phases)
        print(f"Generated {filename}")

    print("Frequency response analysis complete.")


if __name__ == "__main__":
    main()
